### PR TITLE
Fix #1578 - Redesign option parsing for main executable

### LINKED ISSUE #1578: Redesign option parsing for main executable
With #1569, we introduced two additional flags that do not compose with the `--lsp` flag, yet you can pass them all to the executable and nothing complains about that. 
I think we should redesign the options to reflect the two different modes that are currently implemented: `lsp-server` and `project-loading`.
Most important restriction: 
Synopsis and name of the used flags must not change in a backwards incompatible manner.
E.g., the following flags *must* be possible to use together: `--lsp`, `--debug`, `-l log.file` and `--vomit`.

### PR DESCRIPTION
**Pull request by gdziadkiewicz**
### What has been done?
I extracted mode-specific `GlobalOpts` fields into separate types, updated the parsing logic to handle new types as an alternative and updated the places where `GlobalOpts` were used.
### How did I test it?
Added unit tests for option parser exposed by `Haskell.Ide.Engine.Options`.

### CODE REVIEW

**Comment by gdziadkiewicz** [comment]
I will investigate the build failures and mark it as ready to be reviewed after making it build on all envs.

**Comment by gdziadkiewicz** [comment]
@fendor Could you also find a moment to do a review more focused on the style of the code and best practices? I'm willing to learn in the process and to do changes meant to make the new code consistent with the rest of the codebase.

**Comment by alanz** [comment]
As a minimum, please update the description so that when it lands we know what it is about, without having to refer to a (possibly no longer existent) github repo.

**Comment by gdziadkiewicz** [comment]
@alanz I changed the commit message. Please check if it is acceptable now

**Comment by gdziadkiewicz** [comment]
The CI fails don't seem to be connected to this change.

File: src/Haskell/Ide/Engine/Options.hs [hunk 1]
```diff
@@ -14,40 +14,66 @@ import           System.IO
...
+compiler = infoOption hieGhcDisplayVersion
+  (long "compiler" <> help "Show only compiler and version supported")
+
+projectLoadingModeParser :: Parser RunMode
+projectLoadingModeParser =
+  ProjectLoadingMode
+  <$> (ProjectLoadingOpts
+       <$> flag False True
+          (  long "dry-run"
+            <> help "Perform a dry-run of loading files. Only searches for Haskell source files to load. Does nothing if run as LSP server."
```
Comment (fendor) [inline src/Haskell/Ide/Engine/Options.hs]: Line too long, limiting yourself to 80 chars per line makes it easier to read on small monitors

File: test/unit/OptionsSpec.hs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import Haskell.Ide.Engine.Options(GlobalOpts(..), RunMode(..), ProjectLoadingOpts(..), optionParser)
+import System.Exit(ExitCode(..))
+import Data.List(isPrefixOf)
+
+main :: IO ()
+main = hspec spec
+
+spec :: Spec
+spec = do
+  let defaultGlobalOptions = GlobalOpts False Nothing Nothing False Nothing False (ProjectLoadingMode $ ProjectLoadingOpts False [])
```
Comment (fendor) [inline test/unit/OptionsSpec.hs]: Line a bit too long, as well.

**Comment by fendor** [comment]
We will merge next week, when some time has passed since the release

**Comment by fendor** [comment]
We will merge next week, when there was some time since the last release

**Pull request merged**